Reporting API - tidy up CSPViolationReportBody (#35602)
mdn/content: files/en-us/web/api/cspviolationreportbody/sourcefile/index.md

---
title: "CSPViolationReportBody: sourceFile property"
short-title: sourceFile
slug: Web/API/CSPViolationReportBody/sourceFile
page-type: web-api-instance-property
browser-compat: api.CSPViolationReportBody.sourceFile
---

{{APIRef("Reporting API")}}

The **`sourceFile`** read-only property of the {{domxref("CSPViolationReportBody")}} interface indicates the URL of the source file that violated the [Content Security Policy (CSP)](/en-US/docs/Web/HTTP/CSP).

For a violation triggered by the use of an inline script, `sourceFile` is the URL of the current document.
Similarly, if a document successfully loads a script that then violates the document CSP, the `sourceFile` is the URL of the script.
Note however that if a document with a CSP that blocks external resources attempts to load an external resource, `sourceFile` will be `null`.

> This is because the browser extracts the value from _the global object_ of the file that triggered the violation: in this case the external resource is never loaded and therefore has no corresponding global object.

Author: This gave me a lot of grief.

The `sourceFile` is extracted from a global object when processing a trigger. So if you load a script and it does something in violation you have the URL from its global object, and the reported srcFile is the script (not the thing that caused it to be in violation, if you were to say load a file in that script).

However if you have a document and try load a resource in violation of a CSP, there is no srcFile because the trigger did not get to load and so its global object is not available. This weirded me out because the URL is right then in the `src` of the HTML. But there you are.

Because this is loaded from the same domain as the HTML, it is not blocked by the CSP.

The script creates a new {{domxref("ReportingObserver")}} to observe content violation reports of type `"csp-violation"`.
Each time the callback function in invoked, we get the body of the first entry of the reports array, and use it to log the file, line, and column of the violation to the console.

```js
// main.js
const observer = new ReportingObserver(
  (reports, observer) => {
    const cspViolationBody = reports[0].body;
    console.log(`sourceFile: ${cspViolationBody.sourceFile}`);
    console.log(`lineNumber: ${cspViolationBody.lineNumber}`);
    console.log(`columnNumber: ${cspViolationBody.columnNumber}`);
  },
  {
    types: ["csp-violation"],
    buffered: true,
  },
);

observer.observe();
```

Note that while there might be multiple reports in the returned array, for brevity we only log the values of the first element.

#### Results

If serving the above code using a local server (on `http://127.0.0.1:9999/`), the output of the log on Chrome is:

```plain
sourceFile: http://127.0.0.1:9999/inline/
lineNumber: 17

> columnNumber: 0

Author: Note, an error. Hence I report both browser results.

This property is most useful alongside {{domxref("CSPViolationReportBody.lineNumber")}} and {{domxref("CSPViolationReportBody.columnNumber")}}, which provide the location within the file that resulted in a violation.

## Value

A string containing the URL of the file that triggered the violation, or `null`.

## Examples

### CSP inline script violation

This example triggers a CSP violation using an inline script, and reports the violation using a {{domxref("ReportingObserver")}}.

#### HTML

The HTML file below uses the [`<meta>`](/en-US/docs/Web/HTML/Element/meta) element to set the {{httpheader('Content-Security-Policy')}} `default-src` to `self`, which allows scripts and other resources to be loaded from the same domain, but does not allow inline scripts to be executed.
The document also includes an inline script, which should therefore trigger a CSP violation.

```html
<!doctype html>
<html lang="en">
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="default-src 'self'; report-to csp-endpoint" />
    <meta
      http-equiv="Reporting-Endpoints"

>       content="csp-endpoint='https://example.com/csp-reports'" />

Author: @wbamberg In case you are wondering why some examples include `content="default-src 'self'; report-to csp-endpoint" />` unnecessarily, FF appears to need `report-to csp-endpoint` even though it isn't needed in the spec. I should create a bug but have not done so yet.

So the examples don't work without the `report-to`. Note though, I could remove the `Reporting-Endpoints` so report-to points to nothing - that works on FF.

Anyway, I decided that it is generally better just to have the example working if someone copies and pastes.

</html>
```

#### JavaScript (main.js)

The document above also loads the external script `main.js`, which is shown below.
Because this is loaded from the same domain as the HTML, it is not blocked by the CSP.

The script creates a new {{domxref("ReportingObserver")}} to observe content violation reports of type `"csp-violation"`.
Each time the callback function is invoked, we get the body of the first entry of the reports array, and use it to log the file, line, and column of the violation to the console.

```js
// main.js
const observer = new ReportingObserver(
  (reports, observer) => {
    const cspViolationBody = reports[0].body;
    console.log(`sourceFile: ${cspViolationBody.sourceFile}`);
    console.log(`lineNumber: ${cspViolationBody.lineNumber}`);
    console.log(`columnNumber: ${cspViolationBody.columnNumber}`);
  },
  {
    types: ["csp-violation"],
    buffered: true,
  },
);

observer.observe();
```

Note that while there might be multiple reports in the returned array, for brevity we only log the values of the first element.

#### Results

If serving the above code using a local server (on `http://127.0.0.1:9999/`), the output of the log on Chrome is:

```plain

> sourceFile: http://127.0.0.1:9999/inline/

Reviewer: I don't understand why this is `/inline`. Is that the endpoint we are using in the example?

Author: Yes. Arrgg. I already changed fixed this in my test code and later examples (noticed it was confusing when doing the unsafe-inline example). I'll change this everywhere to "test".

Author: ```suggestion
sourceFile: http://127.0.0.1:9999/test/
```

lineNumber: 17
columnNumber: 0
```

Note that the column number is (incorrectly) set to `0`.
The result is similar for Firefox, which correctly reports the `columnNumber`.

```plain

> sourceFile: http://127.0.0.1:9999/inline/

Author: ```suggestion
sourceFile: http://127.0.0.1:9999/test/
```

    console.log(`lineNumber: ${cspViolationBody.lineNumber}`);
    console.log(`columnNumber: ${cspViolationBody.columnNumber}`);
  },
  {
    types: ["csp-violation"],
    buffered: true,
  },
);

observer.observe();
```

Note that while there might be multiple reports in the returned array, for brevity we only log the values of the first element.

#### Results

You can try this out using a [local server](/en-US/docs/Learn/Common_questions/Tools_and_setup/set_up_a_local_testing_server).
Copy the above code into `test/index.html` and `test/main.js` and run the server in the root directory.
Assuming the address of the local server is `http://127.0.0.1:9999`, you can then load the HTML file from `http://127.0.0.1:9999/test/` (or `http://127.0.0.1:9999/test/index.html`).

With the above setup, the output of the log on Chrome is:

```plain
sourceFile: http://127.0.0.1:9999/test/
lineNumber: 15
columnNumber: 0
```

The result is similar for Firefox:

```plain
sourceFile: http://127.0.0.1:9999/test/
lineNumber: 15
columnNumber: 13
```

Note that the column number is different for the two browsers.
Chrome always appears to report `0`.

> The value on Firefox appears to point to the second character after the end of the opening `<script>` element.

Reviewer: ```suggestion
The value on Firefox represents the position of the first character after the end of the opening `<script>` element.
```

---
title: "CSPViolationReportBody: sourceFile property"
short-title: sourceFile
slug: Web/API/CSPViolationReportBody/sourceFile
page-type: web-api-instance-property
browser-compat: api.CSPViolationReportBody.sourceFile
---

{{APIRef("Reporting API")}}

> The **`sourceFile`** read-only property of the {{domxref("CSPViolationReportBody")}} interface indicates the path of the source file that caused the reported CSP violation.

Reviewer: "caused the reported CSP violation" is a great example of the ambiguity in the page at https://pr35602.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CSPViolationReportBody/sample.

In _this_ page (sourceFile), when we say "caused the reported CSP violation", I think we mean, the page that tried to load the blocked resource. e.g. if we have a file "index.html", like:

```html
<!DOCTYPE html>

<html lang="en">
	<head>
		<meta charset="utf-8" />
		<title>example page</title>
		<script src="page.js" defer></script>
	</head>

	<body>
		<script>
			console.log("hello world");
		</script>
	</body>
</html>
```

...and our CSP blocks inline scripts, then I'd expect `sourceFile` to be "index.html".

But in https://pr35602.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/CSPViolationReportBody/sample we say "a string that contains a sample of the resource that caused the [Content Security Policy (CSP)](https://pr35602.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTTP/CSP) violation", but there we mean (I think) it would be part of the inline script: i.e. "the resource that caused the violation" is the blocked resource itself.

Author: The whole thing is horrible. I think I have mostly sorted it with the new wording.
